File: skills/request-z-code/scripts/request_z_code.mjs
#!/usr/bin/env node

import { randomUUID } from "node:crypto";

function configuration() {
  const url = (process.env.ZCODE_ALLOCATOR_URL || "").replace(/\/$/, "");
  const key = process.env.ZCODE_API_KEY || "";
  const agent = (process.env.ZCODE_AGENT_NAME || "").trim().toLowerCase();
  const missing = [
    ["ZCODE_ALLOCATOR_URL", url],
    ["ZCODE_API_KEY", key],
    ["ZCODE_AGENT_NAME", agent],
  ].filter(([, value]) => !value).map(([name]) => name);
  if (missing.length) throw new Error(`Missing required environment values: ${missing.join(", ")}`);
  return { url, key, agent };
}

function parseArgs(argv) {
  const command = argv[0];
  if (!command) throw new Error("A command is required: lookup, allocate, confirm, failed, or status");
  const values = {};
  for (let index = 1; index < argv.length; index += 2) {
    const flag = argv[index];
    if (!flag?.startsWith("--") || argv[index + 1] === undefined) throw new Error(`Invalid argument: ${flag || "<missing>"}`);
    values[flag.slice(2)] = argv[index + 1];
  }
  return { command, values };
}

function requireValues(values, names) {
  const missing = names.filter((name) => !values[name]);
  if (missing.length) throw new Error(`Missing required arguments: ${missing.map((name) => `--${name}`).join(", ")}`);
}

async function apiRequest(method, endpoint, key, body) {
  let lastError;
  for (let attempt = 0; attempt < 2; attempt += 1) {
    try {
      const response = await fetch(endpoint, {
        method,
        headers: {
          Authorization: `Bearer ${key}`,
          Accept: "application/json",
          ...(body ? { "Content-Type": "application/json" } : {}),
        },
        ...(body ? { body: JSON.stringify(body) } : {}),
        signal: AbortSignal.timeout(15000),
      });
      const text = await response.text();
      let result;
      try { result = JSON.parse(text); } catch { result = { error: text }; }
      if (!response.ok) {
        console.error(JSON.stringify({ http_status: response.status, ...result }, null, 2));
        process.exit(2);
      }
      return result;
    } catch (error) {
      lastError = error;
      if (attempt === 0) await new Promise((resolve) => setTimeout(resolve, 1000));
    }
  }
  throw new Error(`Z-Code allocator unavailable: ${lastError?.message || lastError}`);
}

async function main() {
  const { url, key, agent } = configuration();
  const { command, values } = parseArgs(process.argv.slice(2));
  let result;
  if (command === "allocate") {
    requireValues(values, ["name-key", "core", "lane", "page-type"]);
    result = await apiRequest("POST", `${url}/v1/allocate`, key, {
      request_id: values["request-id"] || `${agent}-${randomUUID()}`,
      name_key: values["name-key"],
      z_knowledge_core: values.core.toUpperCase(),
      knowledge_lane: values.lane,
      page_type: values["page-type"],
      requested_by: agent,
    });
  } else if (command === "confirm") {
    requireValues(values, ["z-code", "notion-url"]);
    result = await apiRequest("POST", `${url}/v1/confirm`, key, {
      z_code: values["z-code"], status: "active", notion_url: values["notion-url"],
    });
  } else if (command === "failed") {
    requireValues(values, ["z-code", "reason"]);
    result = await apiRequest("POST", `${url}/v1/confirm`, key, {
      z_code: values["z-code"], status: "failed", reason: values.reason,
    });
  } else if (command === "status") {
    requireValues(values, ["request-id"]);
    result = await apiRequest("GET", `${url}/v1/status/${encodeURIComponent(values["request-id"])}`, key);
  } else if (command === "lookup") {
    requireValues(values, ["name-key"]);
    result = await apiRequest("GET", `${url}/v1/lookup?${new URLSearchParams({ name_key: values["name-key"] })}`, key);
  } else {
    throw new Error(`Unknown command: ${command}`);
  }
  console.log(JSON.stringify(result, null, 2));
}

main().catch((error) => {
  console.error(error.message);
  process.exit(1);
});

File: skills/request-z-code/scripts/request_z_code.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
import uuid


def configuration() -> tuple[str, str, str]:
    url = os.getenv("ZCODE_ALLOCATOR_URL", "").rstrip("/")
    key = os.getenv("ZCODE_API_KEY", "")
    agent = os.getenv("ZCODE_AGENT_NAME", "").strip().lower()
    missing = [name for name, value in (("ZCODE_ALLOCATOR_URL", url), ("ZCODE_API_KEY", key), ("ZCODE_AGENT_NAME", agent)) if not value]
    if missing:
        raise RuntimeError("Missing required environment values: " + ", ".join(missing))
    return url, key, agent


def request(method: str, path: str, key: str, payload: dict | None = None, retries: int = 1) -> dict:
    body = json.dumps(payload).encode("utf-8") if payload is not None else None
    headers = {"Authorization": f"Bearer {key}", "Accept": "application/json"}
    if body is not None:
        headers["Content-Type"] = "application/json"
    for attempt in range(retries + 1):
        try:
            req = urllib.request.Request(path, data=body, headers=headers, method=method)
            with urllib.request.urlopen(req, timeout=15) as response:
                return json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            try:
                parsed = json.loads(detail)
            except json.JSONDecodeError:
                parsed = {"error": detail}
            print(json.dumps({"http_status": exc.code, **parsed}, indent=2), file=sys.stderr)
            raise SystemExit(2) from exc
        except (urllib.error.URLError, TimeoutError) as exc:
            if attempt >= retries:
                raise RuntimeError(f"Z-Code allocator unavailable: {exc}") from exc
            time.sleep(1)
    raise RuntimeError("Z-Code request failed")


def parser() -> argparse.ArgumentParser:
    root = argparse.ArgumentParser(description="Use the authoritative ZedBiz Z-Code allocator")
    sub = root.add_subparsers(dest="command", required=True)
    allocate = sub.add_parser("allocate")
    allocate.add_argument("--request-id", default=None)
    allocate.add_argument("--name-key", required=True)
    allocate.add_argument("--core", required=True)
    allocate.add_argument("--lane", required=True)
    allocate.add_argument("--page-type", required=True)
    confirm = sub.add_parser("confirm")
    confirm.add_argument("--z-code", required=True)
    confirm.add_argument("--notion-url", required=True)
    failed = sub.add_parser("failed")
    failed.add_argument("--z-code", required=True)
    failed.add_argument("--reason", required=True)
    status = sub.add_parser("status")
    status.add_argument("--request-id", required=True)
    lookup = sub.add_parser("lookup")
    lookup.add_argument("--name-key", required=True)
    return root


def main() -> int:
    args = parser().parse_args()
    base_url, key, agent = configuration()
    if args.command == "allocate":
        payload = {
            "request_id": args.request_id or f"{agent}-{uuid.uuid4()}",
            "name_key": args.name_key,
            "z_knowledge_core": args.core.upper(),
            "knowledge_lane": args.lane,
            "page_type": args.page_type,
            "requested_by": agent,
        }
        result = request("POST", f"{base_url}/v1/allocate", key, payload)
    elif args.command == "confirm":
        result = request(
            "POST",
            f"{base_url}/v1/confirm",
            key,
            {"z_code": args.z_code, "status": "active", "notion_url": args.notion_url},
        )
    elif args.command == "failed":
        result = request(
            "POST",
            f"{base_url}/v1/confirm",
            key,
            {"z_code": args.z_code, "status": "failed", "reason": args.reason},
        )
    elif args.command == "status":
        result = request("GET", f"{base_url}/v1/status/{urllib.parse.quote(args.request_id, safe='')}", key)
    else:
        query = urllib.parse.urlencode({"name_key": args.name_key})
        result = request("GET", f"{base_url}/v1/lookup?{query}", key)
    print(json.dumps(result, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

